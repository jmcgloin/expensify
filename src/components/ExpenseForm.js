import React from "react";
import moment from "moment";
import { connect } from "react-redux";
import { SingleDatePicker } from "react-dates";
import { history } from "../routers/AppRouter";

export default class ExpenseForm extends React.Component {
	constructor(props) {
		super(props);
		const edit = !!props.expense;
		this.state = {
			description: edit ? props.expense.description : "",
			amount: edit ? (props.expense.amount/100).toString() : "",
			note: edit ? props.expense.note : "",
			createdAt: edit ? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,
			errorMessage: null
		};
	};
	onDescriptionChange = (e) => {
		const errorMessage = this.state.errorMessage && null;
		this.setState({
			description: e.target.value,
			errorMessage
		});
	};
	onAmountChange = (e) => {
		const amount = e.target.value;
		const errorMessage = this.state.errorMessage && null;
		if(!amount || amount.match(/^\d+(\.\d{0,2})?$/g)) {
			this.setState(() => ({ amount, errorMessage }));
		} else {
			e.target.value = this.state.amount;
		}
	};
	onNoteChange = (e) => {
		this.setState({
			note: e.target.value
		});
	};
	onNewExpenseSubmit = (e) => {
		e.preventDefault();
		if(!this.state.amount || !this.state.description) {
			const errorMessage = "You must enter a description and amount";
			this.setState(() => ({ errorMessage }))
		} else {
			this.state.errorMessage && this.setState(() => ({ errorMessage: null }));
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			});
		}
	};
	onDateChange = (createdAt) => {
		createdAt && this.setState(() => ({ createdAt }));
	};
	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calendarFocused: focused }))
	};
	onCancel = () => {
		history.push("/dashboard");
	}
	render() {
		return (
			<div>
				<h1>Expense Form</h1>
				{this.state.errorMessage}
				<form onSubmit={this.onNewExpenseSubmit}>
					<input 
						type="text"
						name="newDescription"
						placeholder="Enter a new expense"
						autoFocus
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>
					<input
						type="text"
						name="newAmount"
						placeholder="Amount"
						value={this.state.amount}
						onChange={this.onAmountChange}
					/>
					<SingleDatePicker
						date={this.state.createdAt}
						onDateChange={this.onDateChange}
						focused={this.state.calendarFocused}
						onFocusChange={this.onFocusChange}
						numberOfMonths={1}
						isOutsideRange={() => false}
					/>
					<textarea
						name="newNote"
						placeholder="Optional notes" 
						onInput={this.onNoteChange}
					/>
					<button>Submit</button>
					<button onClick={ this.onCancel }>Cancel</button>
				</form>
			</div>
		);
	};
};