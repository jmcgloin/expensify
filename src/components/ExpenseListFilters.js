import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from "../actions/filters";


export class ExpenseListFilters extends React.Component {
	state = {
		calendarFocused: null
	};
	onDatesChange = ({ startDate, endDate }) => {
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	};
	onFocusChange= (calendarFocused) => {
		this.setState(() => ({ calendarFocused }));
	};
	onSortChange = (e) => {
		console.log("value", e.target.value);
		e.target.value === "createdAt" ?
			this.props.sortByDate() :
			this.props.sortByAmount();
	};
	onTextChange = (e) => {
		this.props.setTextFilter(e.target.value);
	};
	render() {
		return (
			<div>
			<input type="text" value={ this.props.filters.text } onChange={ this.onTextChange } />
			<select value={this.props.filters.sortBy} onChange={ this.onSortChange } >
				<option value="createdAt">Date</option>
				<option value="amount">Amount</option>
			</select>
			<DateRangePicker 
				startDate={this.props.filters.startDate}
				startDateId='start'
				endDate={this.props.filters.endDate}
				endDateId='end'
				onDatesChange={this.onDatesChange}
				focusedInput={this.state.calendarFocused}
				onFocusChange={this.onFocusChange}
				numberOfMonths={1}
				isOutsideRange={() => false}
				showClearDates={true}
			/>
		</div>
		)
	}
};

const mapDispatchToProps = (dispatch, props) => {
	return {
		setStartDate: (startDate) => dispatch(setStartDate(startDate)),
		setEndDate: (endDate) => dispatch(setEndDate(endDate)),
		setTextFilter: (text) => dispatch(setTextFilter(text)),
		sortByAmount: () => dispatch(sortByAmount()),
		sortByDate: () => dispatch(sortByDate())
	};
};

const mapStateToProps = (state) => {
	return {
		filters: state.filters
	};
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);