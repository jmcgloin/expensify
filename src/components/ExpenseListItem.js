import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem =  ({ description, id, amount, createdAt, note }) => {
	return (
		<div>
			<h3>{ description }</h3>
			<p>
				Amount: { amount }   |
				Date Created: { createdAt }				
				<Link to={`/edit/${id}`} >
					Edit
				</Link>
			</p>
		</div>
	);
}

export default ExpenseListItem;