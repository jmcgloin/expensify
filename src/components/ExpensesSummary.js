import React from "react";
import numeral from "numeral";
import { connect } from "react-redux";
import getExpensesTotal from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";

export const ExpensesSummary = ({ total, count }) => 
{
	const plural = count === 1 ? "" : "s";
	total = numeral(total / 100).format("$0,0.00");
	return (
		<div>You are currently viewing {count} expense{plural}, totaling {total}</div>
	);
};

const mapStateToProps = (state) => {
	const expenses = selectExpenses(state.expenses, state.filters);
	return {
		total: getExpensesTotal(expenses),
		count: expenses.length
	}
};

export default connect(mapStateToProps)(ExpensesSummary);