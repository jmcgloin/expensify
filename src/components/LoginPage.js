import React from "react";
import { connect } from "react-redux";
import { startLogin } from "../actions/auth";

//project-660304084367

export const LoginPage = ({ startLogin }) => (
	<div>
		<button onClick={ startLogin }>Log In</button>
	</div>
);

const mapDispatchToProps = (dispatch) => ({
	startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);