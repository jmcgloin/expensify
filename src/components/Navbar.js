import React from "react";
import { NavLink } from "react-router-dom";

const Navbar = () => (
	<nav>
		<NavLink to="/" activeClassName="is-active" exact={ true }>Home</NavLink>
		<NavLink to="/create" activeClassName="is-active">Add</NavLink>
		<NavLink to="/help" activeClassName="is-active">Help</NavLink>
		<NavLink to="/dne" activeClassName="is-active">DNE</NavLink>
	</nav>
);

export default Navbar;