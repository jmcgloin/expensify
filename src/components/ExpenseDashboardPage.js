import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";
import { NavLink } from "react-router-dom";

const ExpenseDashboardPage = () => (
	<div>
		<NavLink to="/create" activeClassName="is-active">Add</NavLink>
		<ExpensesSummary />
		<ExpenseListFilters />
		<ExpenseList />
	</div>
);

export default ExpenseDashboardPage;