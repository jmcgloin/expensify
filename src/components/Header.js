import React from "react";
import { connect } from "react-redux";
import { startLogout } from "../actions/auth";
import Navbar from "./Navbar";

export const Header = ({ startLogout }) => (
		<header>
			<h1>Expensify</h1>
			<button onClick={ startLogout }>Log Out</button>
		</header>
);

const mapDispatchToProps = (dispatch) => ({
	startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);