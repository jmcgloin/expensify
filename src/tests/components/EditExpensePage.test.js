import React from "react";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";
import { EditExpensePage } from "../../components/EditExpensePage";

let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
	editExpense = jest.fn();
	removeExpense = jest.fn();
	history = { push: jest.fn() };
	wrapper = shallow(
		<EditExpensePage
		editExpense={ editExpense }
		removeExpense={ removeExpense }
		history={ history }
		expense={ expenses[0] }
	/>);
})

test('should correctly render edit expense page', () => {
	expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense', () => {
	wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
	expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
	expect(history.push).toHaveBeenLastCalledWith('/');
});

test('should handle remove expense', () => {
	wrapper.find('button').simulate('click', { expense: expenses[0]});
	expect(removeExpense).toHaveBeenLastCalledWith(expenses[0].id);
	expect(history.push).toHaveBeenLastCalledWith('/');
});