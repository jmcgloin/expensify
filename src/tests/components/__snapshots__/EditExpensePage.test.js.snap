// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render edit expense page 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 10000,
        "createdAt": "2018-01-01T05:00:00.000Z",
        "description": "A bill",
        "id": 1,
        "note": "A note",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
